java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -491,11 +492,53 @@ public void visitLambdaExpression(LambdaExpressionTree tree) {
         }
       }
       super.visitLambdaExpression(tree);
+      if(lambdaType.isUnknown() || lambdaType.isTagged(JavaType.DEFERRED)) {
+        return;
+      }
+      refineLambdaType(lambdaExpressionTree, lambdaType);
     } else {
       registerType(tree, symbols.deferedType(lambdaExpressionTree));
     }
   }
 
+  private void refineLambdaType(LambdaExpressionTreeImpl lambdaExpressionTree, JavaType lambdaType) {
+    JavaSymbol.MethodJavaSymbol samMethod = getSamMethod(lambdaType);
+    // compute lambda return type
+    JavaType samReturnType = (JavaType) samMethod.returnType().type();
+    JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, lambdaType);
+    // TODO compute return type for lambda with block and return statements
+    if (!lambdaExpressionTree.body().is(Tree.Kind.BLOCK) && !capturedReturnType.is("void")) {
+      Type refinedReturnType = ((AbstractTypedTree) lambdaExpressionTree.body()).symbolType();
+      if (refinedReturnType != capturedReturnType && lambdaType.isParameterized()) {
+        // found a lambda return type different from the one infered : update infered type
+        TypeSubstitution typeSubstitution = ((ParametrizedTypeJavaType) lambdaType).typeSubstitution;
+        if (typeSubstitution.substitutedTypes().contains(capturedReturnType)) {
+          TypeVariableJavaType typeVar = null;
+          for (Map.Entry<TypeVariableJavaType, JavaType> entry : typeSubstitution.substitutionEntries()) {
+            if (entry.getValue() == capturedReturnType) {
+              typeVar = entry.getKey();
+              break;
+            }
+          }
+          if (typeVar != null) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)